File: benchmarks/candidates/libjxl/bench.cpp
// libjxl candidate: decode a JPEG XL file with the raw libjxl C API.
//
// This is the fastest libjxl library path: it talks directly to `JxlDecoder`
// with a resizable thread-pool runner and asks for interleaved float32 RGB,
// which is what an ML / GPU-upload pipeline typically wants. It emits a single
// JSON line matching the jxlit benchmark schema so the compare orchestrator can
// aggregate it alongside every other candidate.

#include <jxl/decode.h>
#include <jxl/resizable_parallel_runner.h>
#include <jxl/types.h>

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <thread>
#include <vector>

namespace {

constexpr int kWarmupDecodes = 3;

struct Args {
  std::string file;
  int iterations = 100;
  int threads = 0;  // 0 == auto
  std::string action = "decode_cpu";
};

// Tolerant parser: understands the flags this candidate needs and accepts (then
// ignores) the rest of the shared benchmark flags so every candidate can be
// launched with an identical argument list.
Args parse_args(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    auto next = [&](const char* name) -> std::string {
      if (i + 1 >= argc) {
        std::fprintf(stderr, "missing value for %s\n", name);
        std::exit(1);
      }
      return argv[++i];
    };
    if (arg == "--file") {
      args.file = next("--file");
    } else if (arg == "--iterations") {
      args.iterations = std::stoi(next("--iterations"));
    } else if (arg == "--threads") {
      args.threads = std::stoi(next("--threads"));
    } else if (arg == "--action") {
      args.action = next("--action");
    } else if (arg == "--layout" || arg == "--hardware" || arg == "--destination") {
      next(arg.c_str());  // accepted for a uniform CLI, not used here
    } else if (arg == "--no-telemetry") {
      // flag with no value; accepted and ignored
    } else {
      std::fprintf(stderr, "unknown argument: %s\n", arg.c_str());
      std::exit(1);
    }
  }
  if (args.file.empty()) {
    std::fprintf(stderr, "--file is required\n");
    std::exit(1);
  }
  if (args.iterations <= 0) {
    std::fprintf(stderr, "--iterations must be greater than 0\n");
    std::exit(1);
  }
  return args;
}

std::vector<uint8_t> read_file(const std::string& path) {
  std::FILE* f = std::fopen(path.c_str(), "rb");
  if (!f) {
    std::fprintf(stderr, "failed to open %s\n", path.c_str());
    std::exit(1);
  }
  std::fseek(f, 0, SEEK_END);
  long size = std::ftell(f);
  std::fseek(f, 0, SEEK_SET);
  std::vector<uint8_t> bytes(static_cast<size_t>(size));
  if (size > 0 && std::fread(bytes.data(), 1, bytes.size(), f) != bytes.size()) {
    std::fprintf(stderr, "failed to read %s\n", path.c_str());
    std::exit(1);
  }
  std::fclose(f);
  return bytes;
}

// Decodes the first frame to interleaved float32 RGB. Returns false on error.
bool decode_once(const uint8_t* data, size_t size, void* runner,
                 std::vector<float>& out, uint32_t& width, uint32_t& height) {
  JxlDecoder* dec = JxlDecoderCreate(nullptr);
  if (!dec) return false;

  bool ok = true;
  if (JxlDecoderSetParallelRunner(dec, JxlResizableParallelRunner, runner) !=
          JXL_DEC_SUCCESS ||
      JxlDecoderSubscribeEvents(dec, JXL_DEC_BASIC_INFO | JXL_DEC_FULL_IMAGE) !=
          JXL_DEC_SUCCESS ||
      JxlDecoderSetInput(dec, data, size) != JXL_DEC_SUCCESS) {
    JxlDecoderDestroy(dec);
    return false;
  }
  JxlDecoderCloseInput(dec);

  const JxlPixelFormat format = {3, JXL_TYPE_FLOAT, JXL_NATIVE_ENDIAN, 0};
  for (;;) {
    JxlDecoderStatus status = JxlDecoderProcessInput(dec);
    if (status == JXL_DEC_ERROR) {
      ok = false;
      break;
    }
    if (status == JXL_DEC_BASIC_INFO) {
      JxlBasicInfo info;
      if (JxlDecoderGetBasicInfo(dec, &info) != JXL_DEC_SUCCESS) {
        ok = false;
        break;
      }
      width = info.xsize;
      height = info.ysize;
      out.resize(static_cast<size_t>(width) * height * 3);
    } else if (status == JXL_DEC_NEED_IMAGE_OUT_BUFFER) {
      size_t buffer_size = 0;
      if (JxlDecoderImageOutBufferSize(dec, &format, &buffer_size) !=
          JXL_DEC_SUCCESS) {
        ok = false;
        break;
      }
      out.resize(buffer_size / sizeof(float));
      if (JxlDecoderSetImageOutBuffer(dec, &format, out.data(), buffer_size) !=
          JXL_DEC_SUCCESS) {
        ok = false;
        break;
      }
    } else if (status == JXL_DEC_FULL_IMAGE) {
      break;  // first frame decoded; that is the image we benchmark
    } else if (status == JXL_DEC_SUCCESS) {
      break;
    } else {
      ok = false;
      break;
    }
  }

  JxlDecoderDestroy(dec);
  return ok;
}

double percentile(const std::vector<double>& sorted, double p) {
  if (sorted.empty()) return 0.0;
  if (sorted.size() == 1) return sorted[0];
  double rank = (p / 100.0) * (sorted.size() - 1);
  size_t lower = static_cast<size_t>(rank);
  size_t upper = std::min(lower + 1, sorted.size() - 1);
  double weight = rank - lower;
  return sorted[lower] * (1.0 - weight) + sorted[upper] * weight;
}

}  // namespace

int main(int argc, char** argv) {
  Args args = parse_args(argc, argv);
  std::vector<uint8_t> bytes = read_file(args.file);

  void* runner = JxlResizableParallelRunnerCreate(nullptr);
  if (!runner) {
    std::fprintf(stderr, "failed to create parallel runner\n");
    return 1;
  }
  uint32_t threads = args.threads > 0
                         ? static_cast<uint32_t>(args.threads)
                         : std::max(1u, std::thread::hardware_concurrency());
  JxlResizableParallelRunnerSetThreads(runner, threads);

  std::vector<float> pixels;
  uint32_t width = 0;
  uint32_t height = 0;
  for (int i = 0; i < kWarmupDecodes; ++i) {
    if (!decode_once(bytes.data(), bytes.size(), runner, pixels, width, height)) {
      std::fprintf(stderr, "warmup decode failed\n");
      JxlResizableParallelRunnerDestroy(runner);
      return 1;
    }
  }

  std::vector<double> latencies_ms;
  latencies_ms.reserve(args.iterations);
  auto decode_start = std::chrono::steady_clock::now();
  for (int i = 0; i < args.iterations; ++i) {
    auto start = std::chrono::steady_clock::now();
    if (!decode_once(bytes.data(), bytes.size(), runner, pixels, width, height)) {
      std::fprintf(stderr, "decode failed\n");
      JxlResizableParallelRunnerDestroy(runner);
      return 1;
    }
    std::chrono::duration<double, std::milli> elapsed =
        std::chrono::steady_clock::now() - start;
    latencies_ms.push_back(elapsed.count());
    // Keep the decoded buffer observable so the decode is not elided.
    volatile float sink = pixels.empty() ? 0.0f : pixels[0];
    (void)sink;
  }
  std::chrono::duration<double> decode_seconds =
      std::chrono::steady_clock::now() - decode_start;

  JxlResizableParallelRunnerDestroy(runner);

  std::vector<double> sorted = latencies_ms;
  std::sort(sorted.begin(), sorted.end());
  double mean = 0.0;
  for (double v : latencies_ms) mean += v;
  mean /= latencies_ms.size();
  double megapixels = static_cast<double>(width) * height / 1e6;

  std::printf(
      "{\"decoder\":\"libjxl\",\"action\":\"%s\",\"hardware\":\"cpu\","
      "\"destination\":\"cpu\",\"iterations\":%d,\"width\":%u,\"height\":%u,"
      "\"channels\":3,\"megapixels\":%.6f,\"decode_seconds\":%.6f,"
      "\"latency_ms\":{\"mean\":%.3f,\"p50\":%.3f,\"p95\":%.3f,\"min\":%.3f,"
      "\"max\":%.3f}}\n",
      args.action.c_str(), args.iterations, width, height, megapixels,
      decode_seconds.count(), mean, percentile(sorted, 50.0),
      percentile(sorted, 95.0), sorted.front(), sorted.back());
  return 0;
}

File: benchmarks/candidates/imagecodecs/bench.py
"""imagecodecs candidate: decode JPEG XL via libjxl and (optionally) upload to GPU.

imagecodecs wraps the libjxl reference decoder, so decode always runs on CPU.
Two destinations are supported:

* ``cpu``: decode into a NumPy array.
* ``gpu``: decode, convert to float32, and upload to a PyTorch device
  (CUDA on Linux, MPS on macOS), synchronizing so the transfer is included in
  the measured time. If no GPU device is available, the worker prints a single
  ``{"skipped": true, ...}`` JSON line and exits 0.

Emits a single JSON line matching the jxlit benchmark schema.
"""

from __future__ import annotations

import argparse
import json
import sys
import time

import imagecodecs
import numpy as np

WARMUP_DECODES = 3


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="imagecodecs decode benchmark")
    parser.add_argument("--file", required=True)
    parser.add_argument("--iterations", type=int, default=100)
    parser.add_argument("--threads", type=int, default=None)
    parser.add_argument("--action", default="decode_cpu")
    parser.add_argument("--destination", choices=("cpu", "gpu"), default="cpu")
    # Accepted for a uniform CLI across candidates, but unused here.
    parser.add_argument("--hardware", default="cpu")
    parser.add_argument("--layout", default="interleaved")
    parser.add_argument("--no-telemetry", action="store_true")
    return parser.parse_args()


def percentile(values: list[float], p: float) -> float:
    if not values:
        return 0.0
    if len(values) == 1:
        return values[0]
    rank = (p / 100.0) * (len(values) - 1)
    lower = int(rank // 1)
    upper = min(lower + 1, len(values) - 1)
    weight = rank - lower
    return values[lower] * (1.0 - weight) + values[upper] * weight


def emit_skip(reason: str) -> None:
    print(json.dumps({"decoder": "imagecodecs", "skipped": True, "reason": reason}))


def pick_torch_device():
    """Return (torch_module, device) or (None, None) if no GPU is available."""
    try:
        import torch
    except ImportError:
        return None, None
    if torch.cuda.is_available():
        return torch, torch.device("cuda")
    mps = getattr(torch.backends, "mps", None)
    if mps is not None and mps.is_available():
        return torch, torch.device("mps")
    return None, None


def synchronize(torch, device) -> None:
    if device.type == "cuda":
        torch.cuda.synchronize()
    elif device.type == "mps":
        torch.mps.synchronize()


def main() -> None:
    args = parse_args()
    with open(args.file, "rb") as handle:
        data = handle.read()

    numthreads = args.threads if args.threads and args.threads > 0 else None

    torch = None
    device = None
    if args.destination == "gpu":
        torch, device = pick_torch_device()
        if device is None:
            emit_skip("no CUDA or MPS device available for GPU destination")
            return

    # Warmup also fixes the output array shape/dtype so the timed loop can decode
    # into a preallocated buffer.
    cpu_out = imagecodecs.jpegxl_decode(data, numthreads=numthreads)
    for _ in range(WARMUP_DECODES - 1):
        imagecodecs.jpegxl_decode(data, numthreads=numthreads, out=cpu_out)
    if device is not None:
        host = np.ascontiguousarray(cpu_out, dtype=np.float32)
        warm = torch.from_numpy(host).to(device)
        synchronize(torch, device)
        del warm

    height = cpu_out.shape[0]
    width = cpu_out.shape[1]
    channels = cpu_out.shape[2] if cpu_out.ndim == 3 else 1
    megapixels = width * height / 1e6

    latencies_ms: list[float] = []
    decode_start = time.perf_counter()
    for _ in range(args.iterations):
        start = time.perf_counter()
        arr = imagecodecs.jpegxl_decode(data, numthreads=numthreads, out=cpu_out)
        if device is not None:
            host = np.ascontiguousarray(arr, dtype=np.float32)
            tensor = torch.from_numpy(host).to(device)
            synchronize(torch, device)
            del tensor
        latencies_ms.append((time.perf_counter() - start) * 1000.0)
    decode_seconds = time.perf_counter() - decode_start

    sorted_ms = sorted(latencies_ms)
    payload = {
        "decoder": "imagecodecs",
        "action": args.action,
        "hardware": "cpu",
        "destination": args.destination,
        "iterations": args.iterations,
        "width": int(width),
        "height": int(height),
        "channels": int(channels),
        "megapixels": megapixels,
        "decode_seconds": decode_seconds,
        "latency_ms": {
            "mean": sum(latencies_ms) / len(latencies_ms),
            "p50": percentile(sorted_ms, 50.0),
            "p95": percentile(sorted_ms, 95.0),
            "min": sorted_ms[0],
            "max": sorted_ms[-1],
        },
    }
    print(json.dumps(payload))


if __name__ == "__main__":
    main()
